pkg/cli/images_sign.go

```diff
@@ -0,0 +1,165 @@
+	cmd := cli.Command(&ImageSign{client: c.ClientFactory}, cobra.Command{
+		Use:               "sign IMAGE_NAME [flags]",
+		Example:           `acorn image sign my-image --key ./my-key`,
+		SilenceUsage:      true,
+		Short:             "Sign an Image",
+		ValidArgsFunction: newCompletion(c.ClientFactory, imagesCompletion(true)).complete,
+		Args:              cobra.ExactArgs(1),
+	})
+	_ = cmd.MarkFlagFilename("key")
+	return cmd
+}
+
+type ImageSign struct {
+	client      ClientFactory
+	Key         string            `usage:"Key to use for signing" short:"k" local:"true" default:"./cosign.key"`
+	Annotations map[string]string `usage:"Annotations to add to the signature" short:"a" local:"true"`
+	Push        bool              `usage:"Push the signature to the signature repository" short:"p" local:"true" default:"true"`
+}
+
+func (a *ImageSign) Run(cmd *cobra.Command, args []string) error {
+	c, err := a.client.CreateDefault()
+	if err != nil {
+		return err
+	}
+
+	targetName := args[0]
+	targetDigest := ""
+
+	img, tag, err := client.FindImage(cmd.Context(), c, args[0])
+	if err != nil && !errors.As(err, &images.ErrImageNotFound{}) {
+		return err
+	}
+
+	if err == nil && tag == "" {
+		return fmt.Errorf("Signing a local image without specifying the repository is not supported")
+	} else if tag != "" {
+		targetName = tag
+		targetDigest = img.Digest
+	}
+
+	ref, err := name.ParseReference(targetName)
+	if err != nil {
+		return err
+	}
+
+	if targetDigest == "" {
+		targetDigest, err = acornsign.SimpleDigest(ref)
+		if err != nil {
+			return err
+		}
+	}
+
+	target := ref.Context().Digest(targetDigest)
+
+	if a.Key == "" {
+		return fmt.Errorf("key is required")
+	}
+	pterm.Info.Printf("Signing Image %s (digest: %s) using key %s\n", targetName, targetDigest, a.Key)
+
+	pass, err := generate.GetPass(false)
+	if err != nil {
+		return err
+	}
+
+	pf := func(_ bool) ([]byte, error) {
+		return pass, nil
+	}
+
+	// Get a sigSigner-verifier from a private key and if the key type is not supported, try to import it first
+	var sigSigner sigsig.SignerVerifier
+	sigSigner, err = signature.SignerVerifierFromKeyRef(cmd.Context(), a.Key, pf) // TODO(iwilltry42): use our own style password prompt
+	if err != nil {
+		if !strings.Contains(err.Error(), "unsupported pem type") {
+			return err
+		}
+		pterm.Debug.Printf("Key %s is not a supported PEM key, importing...\n", a.Key)
+		keyBytes, err := cosign.ImportKeyPair(a.Key, pf)
+		if err != nil {
+			return err
+		}
+		sigSigner, err = cosign.LoadPrivateKey(keyBytes.PrivateBytes, keyBytes.Password())
+		if err != nil {
+			return err
+		}
+	}
+
+	dupeDetector := remote.NewDupeDetector(sigSigner)
+
+	var annotations map[string]interface{}
+	if a.Annotations != nil {
+		annotations = make(map[string]interface{})
+		for k, v := range a.Annotations {
+			annotations[k] = v
+		}
+	}
+
+	payload, signature, err := sigsig.SignImage(sigSigner, target, annotations)
```

Comment:
Is it no OK to just pass `a.Annotations` here directly?

Reply:
Unforatunately not, because down the road we need `map[string]interface{}`, not `map[string]string`.
Alternatively I could make the flag param `map[string]any`, but that would also require new logic in the CLI builder to support that type, so I thought this might be easier.

Reply:
> Alternatively I could make the flag param map[string]any, but that would also require new logic in the CLI builder to support that type

Hmm, I wouldn't have expected new logic because IIUC `any` is just a type alias for `interface{}`.

Reply:
https://goplay.tools/snippet/rrynQWaJ1JV -- but maybe I'm misunderstanding what @iwilltry42 meant by

> I could make the flag param map[string]any

Reply:
@njhale Currently `a.Annotations` is `map[string]string` and I'm doing exactly what you're doing in the playground snippet (converting to `map[string]any` in a loop).

`sigsig.SignImage` accepts `map[string]any`, that's the reason for the conversion.
To use `a.Annotations` directly there it'd need to be `map[string]any` right away, which would require another implementation of `assignMaps` (https://github.com/acorn-io/runtime/blob/main/pkg/cli/builder/builder.go#L288)